File: DataStructuresInC/BinarySearchTree.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node
{
    int item;
    struct Node *left;
    struct Node *right;
} TreeNode;
TreeNode *root = NULL;
void preOrder();
TreeNode *addRecursive(TreeNode *node, int data);
TreeNode *addIterative(TreeNode *node, int data);
int removeRoot(TreeNode *node);
int getMax(TreeNode *node);
int getMin(TreeNode *node);

int main()
{
    root = addRecursive(root, 0);
    addRecursive(root, 1);
    addRecursive(root, 2);
    int value = removeRoot(root);
    preOrder(root);

    return -1;
}
TreeNode *addRecursive(TreeNode *node, int data)
{
    //if the tree ie empty
    if (node == NULL)
    {
        TreeNode *newNode = (TreeNode *)malloc(sizeof(TreeNode));
        newNode->item = data;
        node = newNode;
    }
    //if the data is greater than the parent node value
    else if (node->item > data)
        node->left = addRecursive(node->left, data);
    //if the data is less or equal to the parent node value
    else
        node->right = addRecursive(node->right, data);
    return node;
}
TreeNode *addIterative(TreeNode *node, int data)
{
    TreeNode *newNode = (TreeNode *)malloc(sizeof(TreeNode));
    newNode->item = data;
    node = newNode;
    TreeNode *temp = node;
    TreeNode *curr = NULL;
    while (temp != NULL)
    {
        curr = temp;
        if (data > temp->item)
            temp = temp->left;
        else
            temp = temp->right;
    }
    if (node == NULL)
        temp = newNode;
    else if (data < node->item)
        node->left = newNode;
    else if (data > node->item)
        node->right = newNode;
    return temp;
}
int getMin(TreeNode *node)
{
    int value = 100000;
    TreeNode *temp = node;
    if (node == NULL)
        return value;
    else if (node->left == NULL)
    {
        return node->item;
    }
    else
    {
        while (temp != NULL)
        {
            value = temp->item;
            temp = temp->left;
        }
    }
    return value;
}

int getMax(TreeNode *node)
{
    int value = -1;
    TreeNode *temp = node;
    if (node == NULL)
        return value;
    else if (node->right == NULL)
    {
        return node->item;
    }
    else
    {
        while (temp != NULL)
        {
            value = temp->item;
            temp = temp->right;
        }
    }
    return value;
}

int removeRoot(TreeNode *node)
{
    TreeNode *temp = root;
    if (node == NULL)
        return -1;

    int value = node->item;

    if (node->left == NULL)
        root = (node->right != NULL) ? node->right : NULL;

    if (node->right == NULL)
        root = (node->left != NULL) ? node->left : NULL;

    if (node->left == NULL && node->right == NULL)
        root = NULL;

    else if (node->left != NULL)
    {
        if (root->left->right != NULL)
        {
            temp = temp->left;
            temp->right->right = root->right;
        }
        else
        {
            temp = temp->left;
            temp->right = root->right;
        }
        root = temp;
    }

    return value;
}

void preOrder(TreeNode *node)
{
    if (node == NULL)
        return;
    else
    {
        printf("%d\n", node->item);
        preOrder(node->left);
        preOrder(node->right);
    }
}

File: DataStructuresInJava/BinarySearchTree.java

public class BinarySearchTree {
	Node root;
	public BinarySearchTree() {
		this.root = null;
	}
	
	private class Node{
		Node left;
		Node right;
		int data;
		public Node(int data){
			this.data = data;
			this.right = null;
			this.left = null;
			
		}
	}
	
	public Node addRecursive(Node root, int data) {
		
		Node newNode = new Node(data);
		
		if(root == null ) root = newNode;
		if(data < root.data)  root.left = addRecursive(root.left, data);
		else  root.right = addRecursive(root.right, data);
		this.root = root;
		return this.root;
		
		
	}
	public Node addIterative(Node root, int data) {

		 Node newNode = new Node(data);
		    Node temp = root;
			Node node = null;
			while(temp != null) {
				node = temp;
				if(data < temp.data) temp = temp.left;
				else  temp = temp.right;
			}
			
			if(node == null) 
			temp = newNode;
			else if (data < node.data) node.left = newNode;
			else if (data > node.data) node.right = newNode;
		
		return temp ;
		
	}
	public int getMin(Node node) {
		int value = Integer.MIN_VALUE;
		Node temp = node;
		if(node == null) return value;
		else if(node.left == null) {
			return node.data;
		}
		else {
		while(temp != null) {
			value = temp.data;
			temp = temp.left;
			
		}
		}
		return value;
	}
	public int getMax(Node node) {
		int value = Integer.MAX_VALUE;
		Node temp = node;
		if(node == null) return value;
		else if(node.right == null) {
			return node.data;
		}
		else {
		while(temp != null) {
			value = temp.data;
			temp = temp.right;
			
		}
		}
		return value;
	}
	public int removeRoot(Node root) {
	    Node temp = root;
		if(root == null) return -1;
	    
		int value = root.data;
		
		if(root.left == null) 
			this.root = (root.right != null) ? root.right : null;
			
		if(root.right == null) 
			this.root = (root.left != null) ? root.left : null;
	
		if(root.left == null && root.right == null)  root = null;
			
		else if(root.left != null) {
			if(root.left.right != null) {
				temp = temp.left;
				temp.right.right = root.right;
				
			}
			else {
				temp = temp.left;
				temp.right = root.right;
			}
			this.root = temp;
		}
	
		return value;
	}
	
	public void print(Node node) {
		if(node == null) return;
		
		else {
		
		System.out.println("("+ node.data +")");
		
	    print(node.left);
	   
		print(node.right);
		
		}
	}



	public static void main(String[] args) {
		BinarySearchTree tree = new BinarySearchTree();
		tree.root = tree.addIterative(tree.root, 8);
		tree.addIterative(tree.root, 6);
		tree.addIterative(tree.root, 5);
		tree.addIterative(tree.root, 7);
		tree.addIterative(tree.root, 10);
		tree.addIterative(tree.root, 9);
		tree.addIterative(tree.root, 11);
		System.out.println("Min "+tree.getMin(tree.root));
		System.out.println("Max  "+tree.getMax(tree.root));
	
		tree.print(tree.root);
	//	tree.print(tree.root);
	//	tree.removeRoot(tree.root);
	//	tree.print(tree.root);
	//	
	    
	   
	  
	  

	}

}
